### PR TITLE
Introduce deprecation categories

### LINKED ISSUE #64824: Should we include a category with deprecation messages?
Right now, the deprecation messages that Elasticsearch records contain a "key" (a string identifier, loosely descriptive) and a message about the deprecation, possibly with specific details.

We discussed during an 8.0 Stack Upgrade meeting whether it would be useful to also record a category field, e.g. "search" or "security". This would provide a useful field for grouping messages in the Kibana Upgrade Assistant. Such groupings would be useful for prioritising efforts to address deprecated uses of Elasticsearch, or even assigning work to different groups.

The work to implement this would be simple enough, just a little tedious.

### PR DESCRIPTION
**Pull request by pugnascotia**
Closes #64824. Introduce the concept of categories to deprecation logging. Every location where we log a deprecation message must now include a deprecation category.

See also #67266.

All opinions welcome on the categories I've used here. I'd be very happy to shrink the number here, but I also don't want the categories to be so broad as to be meaningless.

Note that I probably won't directly backport this - instead, I expect I'll have to reapply the `DeprecationMessage` change and then go through all the call sites again.

### CODE REVIEW

**Review by jdconrad — COMMENTED** [review]
@pugnascotia I wonder if we should add more categories such as "plugin" or "parsing" since those would cover a lot of the "other" category if we consider most of the date deprecations as parsing issues.

File: server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java [hunk 1]
```diff
@@ -86,14 +86,19 @@ private static String toLoggerName(final Class<?> cls) {
      * Logs a message at the {@link #DEPRECATION} level. The message is also sent to the header warning logger,
      * so that it can be returned to the client.
      */
-    public DeprecationLoggerBuilder deprecate(final String key, final String msg, final Object... params) {
-        return new DeprecationLoggerBuilder().withDeprecation(key, msg, params);
+    public DeprecationLoggerBuilder deprecate(
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java:89]: btw do you think it would be worthy to make this method work in a "fluent" way?
I initialliy thought that deprecation logger could be used like
```
deprecationLogger.deprecate("msg")
.compatibleWarning("msg2")
```
The compatible logger is still not merged (going to be soon I hope), but maybe we could use the builder here?
I am worried about the performance and we won't be creating too much objects here though..
WDYT?

**Review by pgomulka — APPROVED** [review]
I left one comment, but it looks good to me

File: server/src/main/java/org/elasticsearch/common/logging/DeprecationCategory.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.logging;
+
+/**
+ * Deprecation log messages are categorised so that consumers of the logs can easily aggregate them.
+ */
+public enum DeprecationCategory {
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/common/logging/DeprecationCategory.java:25]: would would it make any sense for this to be an interface and some of the categories to be defined in server and some in plugins/xpack?
We will be using deprecations a lot in `x-pack/rest-compatible` plugin and I wonder if we should be adding all the categories from plugins here?

Comment (pugnascotia) [reply server/src/main/java/org/elasticsearch/common/logging/DeprecationCategory.java:25]: I'm not sure about doing that, as the categories ought to be reasonable general and not too fine-grained. We could change the implementation later if we feel that we do need more flexibility.

File: server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java (hunk 1 above)
Comment (pugnascotia) [reply server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java:89]: I don't follow - `deprecationLogger.deprecate(...)` already returns a `DeprecationLoggerBuilder`, so isn't it already fluent?

**Comment by pugnascotia** [comment]
Thanks @jdconrad, I've added the categories that you suggested.

@rjernst are you at all interested in checking over this?

File: server/src/main/java/org/elasticsearch/common/logging/DeprecatedMessage.java [hunk 3]
```diff
@@ -31,13 +33,14 @@
...
-    public static ESLogMessage of(String key, String xOpaqueId, String messagePattern, Object... args) {
+    public static ESLogMessage of(DeprecationCategory category, String key, String xOpaqueId, String messagePattern, Object... args) {
         ESLogMessage esLogMessage = new ESLogMessage(messagePattern, args)
             .field("data_stream.type", "logs")
             .field("data_stream.dataset", "deprecation.elasticsearch")
             .field("data_stream.namespace", "default")
             .field("ecs.version", ECS_VERSION)
-            .field("key", key);
+            .field("key", key)
+            .field("elasticsearch.event.category", category.name().toLowerCase(Locale.ROOT));
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/logging/DeprecatedMessage.java:43]: Why not just `category`? Is this ECS related?

**Review by rjernst — APPROVED** [review]
LGTM, one minor question

File: server/src/main/java/org/elasticsearch/common/logging/DeprecatedMessage.java (hunk 3 above)
Comment (pugnascotia) [reply server/src/main/java/org/elasticsearch/common/logging/DeprecatedMessage.java:43]: Yes, the broader change is in #67266.
